### DISCUSSION TITLE
SegManualMaskPredictor().image_predict() not returning masks?

### DISCUSSION THREAD

**Discussion by saschwarz** [kadirnar/segment-anything-video#52] [General]
It is likely I'm doing something wrong, any help is greatly appreciated.

It looks like `SegManualMaskPredictor().image_predict()` is correctly generating and displaying the input bounding box and has generated an annotation within that bounding box. But, the masks returned don't seem to have the annotation in it?

Here is a link to a collab notebook that reproduces it: 
https://colab.research.google.com/github/saschwarz/yolov8-bbox-segment-anything/blob/main/Yolov8_bbox_segment_anything.ipynb

<img width="1607" alt="Screenshot 2023-04-14 at 6 49 46 AM" src="https://user-images.githubusercontent.com/184171/232051982-57989778-7a0b-4b3a-ac4c-c5ea92788d9c.png">

**Comment by kadirnar** [comment]
I will fix this problem. Can you use this code for now?

```python
from metaseg import sahi_sliced_predict, SahiAutoSegmentation

bbox =         # yolov8 bbox
image_path = "test.png"
SahiAutoSegmentation().predict(
    source=image_path,
    model_type="vit_b",
    input_box=boxes,
    multimask_output=False,
    random_color=False,
    show=True,
    save=False,
)
```
You can also use the SAHI library to use the Yolov8 model.
https://github.com/obss/sahi/blob/main/demo/inference_for_yolov8.ipynb

**Reply by saschwarz** [reply]
Thanks for the fast reply! I've tried with a yolov8 bounding box that has values 0<= x <=1 and 0<= y <=1 and also using pixel values (e.g `[167, 218, 47, 137]`). In the first case no masks are found and in the second case I get a traceback:

```
jumps-3_277225_png.rf.a788d4e02595959f5f497670dab7d536.jpg 1599 2046 73
[167, 218, 47, 137]
vit_l model already exists as 'vit_l.pth'. Skipping download.
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-29-230668ceaa36>](https://localhost:8080/#) in <cell line: 18>()
     24   input_boxes = [[int(b.xmid), int(b.ymid), int(b.width), int(b.height)] for b in annotation.boxes]
     25   print(input_boxes[0])
---> 26   result = SahiAutoSegmentation().predict(
     27       source=full_path,
     28       model_type="vit_l",

[/usr/local/lib/python3.9/dist-packages/metaseg/sahi_predict.py](https://localhost:8080/#) in predict(self, source, model_type, input_box, input_point, input_label, multimask_output, random_color, show, save)
    108         plt.imshow(read_image)
    109         for mask in masks:
--> 110             plt_load_mask(mask.cpu().numpy(), plt.gca(), random_color=random_color)
    111         for box in input_boxes:
    112             plt_load_box(box.cpu().numpy(), plt.gca())

AttributeError: 'numpy.ndarray' object has no attribute 'cpu'
```

If it helps I've updated my notebook and the code is in the final cell.

**Comment by kadirnar** [comment]
Can you write the output of the input_boxes variable? Can you try input_boxes instead of input_boxes[0]?

**Reply by saschwarz** [reply]
Input_boxes in that example is `[167, 218, 47, 137]`

**Reply by saschwarz** [reply]
If I give a list of more than one input_box the collab backend crashes. So I gave it a list of a single bounding box. 

But, I don't think it is plotting the bbox correctly or I'm not supplying it correctly. It should be [center_x, center_y, width, height] in pixels?

<img width="940" alt="Screenshot 2023-04-14 at 10 50 40 AM" src="https://user-images.githubusercontent.com/184171/232093491-71d30e78-71cf-45e7-b7a2-7c85730c1623.png">

**Reply by saschwarz** [reply]
Ok, did some more testing, it looks like it is expecting bbox to be in pixels [xmin, ymin, xmax, ymax]. 
The result returned is an image and not the segment annotations(?)

<img width="624" alt="Screenshot 2023-04-14 at 11 32 32 AM" src="https://user-images.githubusercontent.com/184171/232103579-8a10fe6b-f8f8-4394-adde-9d588d259960.png">

**Comment by saschwarz** [comment]
I tried changing `save=False` and `show=False` and now the method returns a `torch.Tensor` but in the same format as calling the original predictor. Interestingly, it still shows the image too.

<img width="879" alt="Screenshot 2023-04-14 at 11 35 59 AM" src="https://user-images.githubusercontent.com/184171/232104641-7795b1ee-73cd-42c8-9bea-a6d4683d8c87.png">

**Reply by kadirnar** [reply]
Did you restart kernel?

**Reply by saschwarz** [reply]
Yes, all these results are after I restarted/re-imported dependencies.

So four possible things: SegManualMaskPredictor().image_predict() and SahiAutoSegmentation.predict() return a Tensor instead of the payload in the docstring. Also, SahiAutoSegmentation doesn't seem to do what is expected for save/show kwargs and also no output_path keyword accepted by SahiAutoSegmentation.

Thanks.

**Reply by kadirnar** [reply]
I will publish colab notebook for demo work. I'm very busy right now, this might take a while.

**Reply by kadirnar** [reply]
https://github.com/kadirnar/segment-anything-video/issues/53

**Reply by saschwarz** [reply]
I appreciate you help!

I can't get all the packages installed locally, if I can, I can try to debug it, though all these packages are new to me.

Honestly, I just need to get the annotations out of either of these predictors, I don't care about any of the other issues.

**Reply by kadirnar** [reply]
Why can't you install all packages? You just want polygan coordinates?

**Reply by saschwarz** [reply]
I can't get pybedtools installed on my Mac using pip. 

Yes, I'm trying to use MetaSeg to convert 50 yolov8 images annotated with bounding boxes to be annotated with polygons (because I need to calculate center and rotation with my model). So if either predictor returned the polygons I'd be almost done 😄
